paimon-core/src/main/java/org/apache/paimon/mergetree/compact/MergeTreeCompactRewriter.java

```diff
@@ -104,7 +104,9 @@ protected CompactResult rewriteCompaction(
 
         List<DataFileMeta> before = extractFilesFromSections(sections);
         notifyRewriteCompactBefore(before);
-        return new CompactResult(before, writer.result());
+        List<DataFileMeta> after = writer.result();
+        after = notifyRewriteCompactAfter(after);
+        return new CompactResult(before, after);
```

Comment:
1. For files in `before`, you also need to delete their lookup files. Add tests on snapshot expire to check this.
2. No need to create another lookup file, if the original file is upgraded, but not compacted.